extmod/modselect.c

```diff
@@ -26,62 +26,223 @@
+    poll_obj->flags_ret = flags_ret;
+}
+
+#endif
+
+STATIC void poll_group_add(poll_group_t *poll_group, const mp_obj_t *obj, mp_uint_t obj_len, mp_uint_t flags, bool or_flags) {
     for (mp_uint_t i = 0; i < obj_len; i++) {
-        mp_map_elem_t *elem = mp_map_lookup(poll_map, mp_obj_id(obj[i]), MP_MAP_LOOKUP_ADD_IF_NOT_FOUND);
+        mp_map_elem_t *elem = mp_map_lookup(&poll_group->map, mp_obj_id(obj[i]), MP_MAP_LOOKUP_ADD_IF_NOT_FOUND);
         if (elem->value == MP_OBJ_NULL) {
             // object not found; get its ioctl and add it to the poll list
-            const mp_stream_p_t *stream_p = mp_get_stream_raise(obj[i], MP_STREAM_OP_IOCTL);
             poll_obj_t *poll_obj = m_new_obj(poll_obj_t);
             poll_obj->obj = obj[i];
+
+            #if MICROPY_PY_SELECT_POSIX_OPTIMISATIONS
+            int fd = -1;
+            if (mp_obj_is_int(obj[i])) {
+                // A file descriptor integer passed in as the object, so use it directly.
+                fd = mp_obj_get_int(obj[i]);
+                if (fd < 0) {
+                    mp_raise_ValueError(NULL);
+                }
+                poll_obj->ioctl = NULL;
+            } else {
+                // An object passed in.  Check if it has a file descriptor.
+                const mp_stream_p_t *stream_p = mp_get_stream_raise(obj[i], MP_STREAM_OP_IOCTL);
+                poll_obj->ioctl = stream_p->ioctl;
+                int err;
+                mp_uint_t res = stream_p->ioctl(obj[i], MP_STREAM_GET_FILENO, 0, &err);
+                if (res != MP_STREAM_ERROR) {
+                    fd = res;
+                }
+            }
+            if (fd >= 0) {
+                // Object has a file descriptor so add it to pollfds.
+                poll_obj->pollfd = poll_group_add_fd(poll_group, fd);
+            } else {
+                // Object doesn't have a file descriptor.
+                poll_obj->pollfd = NULL;
+            }
+            #else
+            const mp_stream_p_t *stream_p = mp_get_stream_raise(obj[i], MP_STREAM_OP_IOCTL);
             poll_obj->ioctl = stream_p->ioctl;
-            poll_obj->flags = flags;
-            poll_obj->flags_ret = 0;
+            #endif
+
+            poll_obj_set_flags(poll_obj, flags);
+            poll_obj_set_flags_ret(poll_obj, 0);
             elem->value = MP_OBJ_FROM_PTR(poll_obj);
         } else {
             // object exists; update its flags
+            poll_obj_t *poll_obj = (poll_obj_t *)MP_OBJ_TO_PTR(elem->value);
             if (or_flags) {
-                ((poll_obj_t *)MP_OBJ_TO_PTR(elem->value))->flags |= flags;
+                poll_obj_set_flags(poll_obj, poll_obj_get_flags(poll_obj) | flags);
             } else {
-                ((poll_obj_t *)MP_OBJ_TO_PTR(elem->value))->flags = flags;
+                poll_obj_set_flags(poll_obj, flags);
             }
         }
     }
 }
 
 // poll each object in the map
```

Comment:
Unless you know (from knowing about this PR) what's going on here, the concept of a "single round" might not be that obvious. The commit description doesn't explain it either.

I'd probably suggest renaming `poll_group_poll_all` to `poll_group_poll_until_ready_or_timeout` and then perhaps this function should be `poll_group_poll_once`.

Reply:
Renamed as suggested.